cpython: fix cross-compiling extension modules

lopsided98 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Fixes cross-compiling Python native extension modules to host platforms with a different word length than the build platform as well as in a few other cases.

Currently, cross-compiling Python extension modules generally works assuming the build and host platforms have the same word size. This leads to problems like those seen here: https://github.com/NixOS/nixpkgs/issues/50925#issuecomment-648771961, where x86_64 -> aarch64 works, but x86_64 -> armv7l doesn't. Unfortunately, this PR doesn't fix `btrfs-progs` (https://github.com/NixOS/nixpkgs/issues/50925) and `talloc` (https://github.com/NixOS/nixpkgs/issues/63043), because they do not use setuptools/distutils.

This happens because the build Python's headers are used, rather than host Python's headers. The header location and other data used for compiling extension modules are available in the `_sysconfigdata__<kernel>_<platform>` module. `distutils.sysconfig` can be patched to use this data and then Python needs to be told to use `sysconfigdata` from host Python using the `_PYTHON_SYSCONFIGDATA_NAME` environment variable. I believe the patch should be upstreamable.

There was an ill-fated upstream PR to improve and document Python cross-compilation, avoiding the need to manually construct `_PYTHON_SYSCONFIGDATA_NAME`, but based on my quick review I don't think it would have worked for us as is: https://github.com/python/cpython/pull/17420

Setting `_PYTHON_HOST_PLATFORM` and `_PYTHON_SYSCONFIGDATA_NAME` appropriately also fixes some other issues, such as adding the correct platform suffix to extension module file names. This allows `cryptography` to build after a few minor fixes.

I have tested this PR by cross-compiling a number of Python packages for armv6l, including `requests` and `netifaces`, both of which failed before.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lopsided98 committed:
cpython: fix finding headers when cross-compiling extension modules

lopsided98 commented:
cc @FRidh @Ericson2314

lopsided98 commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -179,7 +176,25 @@ let
+    # Python packages built through cross-compilation are always for the host platform.
+    disallowedReferences = [ python.pythonForBuild ];
+    # Make distutils compile extension modules against host Python
+    _PYTHON_HOST_PLATFORM = pythonHostPlatform;
```
It was suggested in https://github.com/NixOS/nixpkgs/issues/53320#issuecomment-572982797 to set this in the setup hook. I'm not sure which is more appropriate.

FRidh replied:
This is best put in `pkgs/development/interpreters/python/setup-hook.sh`. That contains interpreter-specific configuration when using the interpreter in a build.

lopsided98 replied:
I'm working on adding this to the setup hook, but these variables need to only be set for host Python and I can't think of a way to detect that in the setup hook. Setting them if Python is cross-compiled would probably work in practice, but that seems kind of hacky.

lopsided98 commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -179,7 +176,25 @@ let
+    _PYTHON_SYSCONFIGDATA_NAME = "_sysconfigdata__${parsed.kernel.name}_${pythonHostPlatform}";
+    # Allow host Python's sysconfigdata to be imported
+    preBuild = ''
+      export PYTHONPATH="${python}/lib/${python.libPrefix}:$PYTHONPATH"
```
Again, maybe in a setup hook. Normally this path is implicitly added to `sys.path`, but this is needed here because we are mixing build and host libraries.

FRidh replied:
These indeed belong in the setup hook.

But, adding the host `lib/pythonx.y` to `PYTHONPATH` seems tricky, because it also includes other native modules. Would it be wiser to copy the `_sysconfigdata__*.py` to another derivation/output and adding only that one?

lopsided98 replied:
I think it is safe to add this path to `PYTHONPATH` because all the native modules have platform specific suffixes, which cpython seems to use to decide which to load. That's how `cryptography` is able to work with both host and build `cffi` on `PYTHONPATH`.

lopsided98 replied:
I found a problem with this: the `subprocess` module hardcodes a path to bash, and if this comes first it tries to invoke host bash. Moving it to the end of `PYTHONPATH` should probably fix it.

lopsided98 replied:
Actually, that's wrong because `PYTHONPATH` is prepended to `sys.path`. I guess we need to add sysconfigdata to its own derivation like you suggested.

lopsided98 commented on pkgs/development/python-modules/cryptography/default.nix:
```diff
@@ -31,6 +32,7 @@ buildPythonPackage rec {
 
   outputs = [ "out" "dev" ];
 
+  nativeBuildInputs = stdenv.lib.optional (!isPyPy) python.pythonForBuild.pkgs.cffi;
```
Splicing doesn't seem to work here. Anyone know why?

FRidh replied:
https://github.com/NixOS/nixpkgs/issues/49526

FRidh replied:
If this is always the case we are going to have a bit of a problem, because all the `nativeBuildInputs` in case of Python packages are then incorrect.

lopsided98 commented:
I created an upstream PR: https://github.com/python/cpython/pull/22440

lopsided98 commented:
I have now implemented this PR as a setup hook. I did not add it to the common Python `setup-hook.sh` because the environment variables are cpython specific and because I needed access to cpython's `$out`, which cannot be done easily if the setup hook is defined in a separate derivation. To get the sysconfigdata module on `PYTHONPATH`, I moved it to `site-packages`.

Also, this PR now fixes #50925.

lopsided98 commented:
After looking into it some more, this PR also helps solve #63043. Normally talloc uses `python-config`, but it will fall back to using the `sysconfig` module if `python-config` is not found ([apparently this is to support Windows](https://github.com/samba-team/samba/blob/0b742ec6a0558397d5cf01b99a401f8e2bc0e2e0/third_party/waf/waflib/Tools/python.py#L387), but it seems to work fine on Linux). You can force this to happen by setting `PYTHON_CONFIG=/invalid` in the environment. This also applies to samba and tdb, which all use the same build system.

Thra11 commented:
This doesn't fix #63043 for me. Running `nix-build '<nixpkgs>' -A pkgsCross.armv7l-hf-multiplatform.talloc` I still get:
```
Checking if compiler accepts -fstack-clash-protection                             : yes
Checking linker accepts -Wl,-no-undefined                                         : yes
Checking linker accepts ['-undefined', 'dynamic_lookup']                          : no
Checking for program 'python3'                                                    : /nix/store/0w5g6pmsb0i7f7vklbg2snyj3wfalq7y-python3-3.8.5/bin/python3
Checking for program 'python'                                                     : /nix/store/0w5g6pmsb0i7f7vklbg2snyj3wfalq7y-python3-3.8.5/bin/python3
Checking for program 'python3'                                                    : /nix/store/0w5g6pmsb0i7f7vklbg2snyj3wfalq7y-python3-3.8.5/bin/python3
Checking for python version >= 3.5.0                                              : 3.8.5
python-config                                                                     : /nix/store/0w5g6pmsb0i7f7vklbg2snyj3wfalq7y-python3-3.8.5/bin/python3-config
Asking python-config for pyembed '--cflags --libs --ldflags --embed' flags        : yes
Testing pyembed configuration                                                     : Could not build a python embedded interpreter
Testing pyembed configuration                                                     : Could not build a python embedded interpreter
The configuration failed
(complete log in /build/talloc-2.3.1/bin/config.log)
builder for '/nix/store/5bn9ss7a1al5js5l0ncz0n4q5i3hqgc9-talloc-2.3.1-armv7l-unknown-linux-gnueabihf.drv' failed with exit code 1
```

Thra11 commented:
Unfortunately, `--keep-failed` doesn't seem to work any more, so I can't check `config.log`.

lopsided98 commented:
Yes, it doesn't solve #63043 by itself, so also need to set `PYTHON_CONFIG="/invalid"` in talloc.

Thra11 commented:
Ah. Missed that bit, sorry!

Thra11 commented:
Still doesn't quite work for me:
```
Checking for python version >= 3.5.0                                              : 3.8.5
python-config                                                                     : /invalid
Checking for library python3.8 in LIBPATH_PYEMBED                                 : not found
Checking for library python3.8 in LIBDIR                                          : not found
Checking for library python3.8 in python_LIBPL                                    : yes
Checking for header Python.h                                                      : Distutils not installed? Broken python installation? Get python-config now!
The configuration failed
(complete log in /build/talloc-2.3.1/bin/config.log)
builder for '/nix/store/8jf92la0x7x00wfaffka361s1qipggcb-talloc-2.3.1-armv7l-unknown-linux-gnueabihf.drv' failed with exit code 1
```

lopsided98 commented:
Oh right, you also have to add both host and build Python as dependencies. I've added the full fix to this PR.

Thra11 commented:
Thanks! I've confirmed that it does indeed work.

FRidh commented on pkgs/development/python-modules/cryptography/default.nix:
```diff
@@ -52,6 +54,9 @@ buildPythonPackage rec {
   '';
 
+  # False positive due to both build and host cffi on PYTHONPATH
+  dontUsePythonCatchConflicts = true;
```
How should Python path looking when cross-compiling a package that relies on native code, both `host` and `build`? We need to find a solution for this so the hook implementing `dontUsePythonCatchConflicts` functions properly in that case as well.

lopsided98 replied:
I guess we should make an exception for host and build versions of the same package, but I'm not sure how to implement that. The current hook uses `pkg_resources`, and the `pkg_resources.Distribution` object has a `platform` attribute which seems like it is supposed to do what we want:

> A string representing the platform the distribution is intended for, or None if the distribution is “pure Python” and therefore cross-platform. See Platform Utilities below for more information on platform strings.

This attribute is determined by parsing the name of the `.dist-info` directory, and unfortunately all the packages I have checked don't have a platform string.

FRidh replied:
We probably need two different `PYTHONPATH`s. Until that is solved, we could exclude this hook (see `mk-python-derivation.nix` when cross-compiling.

FRidh replied:
Disabled the check for now in 27d3a18ac7d34b66a14b634cb785e8c8caef024b.

FRidh commented on pkgs/development/interpreters/python/cpython/default.nix:
```diff
@@ -88,6 +88,33 @@ let
     "$out/bin/python"
   else pythonForBuild.interpreter;
 
+  sysconfigdataHook = with stdenv.hostPlatform; let
```
I'd appreciate a comment here explaining briefly what sysconfig is, what this hook does and why.

FRidh commented on pkgs/development/libraries/talloc/default.nix:
```diff
@@ -56,4 +57,9 @@ stdenv.mkDerivation rec {
   };
-}
+} // stdenv.lib.optionalAttrs (stdenv.hostPlatform != stdenv.buildPlatform) {
+  # python-config from build Python gives incorrect values when cross-compiling.
```
Is this likely to occur more often? If so, it should probably be part of a hook.

lopsided98 replied:
It occurs for all samba packages that use Python: `ldb`, `talloc`, `tdb`, `tevent`, `samba` and possibly `ntdb` (it fails for other reasons that I haven't looked into). Is that enough to justify a hook?

FRidh commented:
Related PR https://github.com/NixOS/nixpkgs/pull/104201.

FRidh commented:
Pushed 11806b6ede420bbdb980ca1c36918b71e9ddc0a2 and 6501d9880c3f612afd417e7c45fff2f58f19009a and was able to cross-compile cffi and numpy. Thanks!

FRidh commented:
Also pushed 489912ee8b863b6f66aae8cf924c5fb03a0f1244 for cryptography.

FRidh commented:
Reverted, made a fix, and now reverted again in 0a12b8a5e9f31ab1f56f55df93af24b46dd0b30f. This is going to need significantly more testing.

All Python versions need to build Python packages: native, cross, i686.

lopsided98 commented:
I've reopened this PR and pushed a updated commit that should fix the i686 build. I'm currently building the tests.

Ericson2314 commented:
I pushed `master` to `staging` because @FRidh and I today did some stuff that might be relevant.

lopsided98 commented:
@ofborg build python3.tests

lopsided98 commented:
I fixed the older Python versions as well now.

@ofborg build python36.tests python37.tests python38.tests python39.tests python310.tests

lopsided98 commented:
3.10 was already broken.

@ofborg build python36.tests, python37.tests, python38.tests, python39.tests

FRidh commented:
Currently we're working on updating the Python packages set. When that is done, I'll push this job to that branch for further testing. That will be about a week from now.

Atemu commented:
Hi, it's been two weeks; any update?

flokli commented:
I can confirm this fixes https://github.com/NixOS/nixpkgs/issues/107578.

I cherry-picked this on top of currently master locally and will run some more tests.

@FRidh is this pushed somewhere to a jobset?

flokli commented:
I pushed the cherry-pick to `cpython-headers-cross`.

This reproducibly breaks the test of `libuv` (on x86_64-linux), which is a dependency of cmake:

```
not ok 72 - fs_fstat
# exit code 134
# Output from process `fs_fstat`:
# Assertion failed in test/test-fs.c on line 1389: s->st_dev == (uint64_t) t.st_dev
```

`pkgs.pkgsCross.raspberryPi.libuv` builds, however.

flokli commented:
I was able to successfully build `-A python36.tests -A python37.tests -A python38.tests -A python39.tests`.

`python310.tests` failed before and after this.

FRidh commented:
Pushed 095955212cc71b993378e0102f682b9165f376e7 to `python-unstable`. Should make it easier to test wider.

FRidh commented:
https://hydra.nixos.org/eval/1637790 looks good. This now needs a rebase and then it can go in staging.

flokli commented:
I rebased this to latest staging and pushed again.

The only conflict was the addition of the `# Backport a fix for ctypes.util.find_library.` patch in `pkgs/development/interpreters/python/cpython/default.nix` in staging, which was causing git to not find the exact place to add the patches introduced here.

flokli commented:
I successfully built all of `python3{6,7,8,9}.tests` and confirmed this still fixes `pkgsCross.raspberryPi.libselinux`, so it doesn't look like there's any conflicts due to other changes in staging in the meantime. Let's get this in :-)

matthewbauer commented:
Would it be possible to apply this to 20.09? I think we may be able to avoid the mass-rebuild...

jonringer commented:
as long it doesn't have any breaking behavior, I'm fine with backporting it

Ericson2314 commented:
@matthewbauer While the first version of this was made independently, it's possible the final version relies on https://github.com/NixOS/nixpkgs/pull/104201 in some ways, too.

FRidh commented:
A backport would need extensive testing as well first.

On Tue, 29 Dec 2020, 01:53 Matthew Bauer, <notifications@github.com> wrote:

> Would it be possible to apply this to 20.09? I think we may be able to
> avoid the mass-rebuild...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/98915#issuecomment-751909106>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQHZ34IC74C5UOK37MP4GDSXESATANCNFSM4R3KRSSA>
> .
>

jonringer commented:
> A backport would need extensive testing as well first.

This is my biggest concern. I would prefer not to backport if the scope of everything needed to backport is ill-defined, and likely to introduce regressions. Unstable works great for iteration, but the purpose of the release channel is to allow people to update without fear.

pjjw commented:
hey just FYI i think this PR broke python extensions under musl libc, have a PR out to fix at #110293. looks like you all are talking about backporting this and it'd be a bummer to backport that breakage too.
